src/build_options.rs

```diff
@@ -725,6 +691,166 @@ impl BuildOptions {
+
+            #[allow(clippy::comparison_chain)]
+            if platform_tags.len() > 2 {
+                bail!(
+                    "Expected only one or two platform tags but found {}",
+                    platform_tags.len()
+                );
+            } else if platform_tags.len() == 2 {
+                // The two platform tags can't be the same kind
+                let tag_types = platform_tags
+                    .iter()
+                    .map(|tag| tag.is_musllinux())
+                    .collect::<HashSet<_>>();
+                if tag_types.len() == 1 {
+                    bail!(
+                        "Expected only one platform tag but found {}",
+                        platform_tags.len()
+                    );
+                }
+            }
+        }
+        _ => {
+            if platform_tags.len() > 1 {
+                bail!(
+                    "Expected only one platform tag but found {}",
+                    platform_tags.len()
+                );
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Returns the merged list of dependencies
+///
+/// Ideally, we'd like to parse the version constraints from the user and merge them with those
+/// from the transitive requirements, but that's not implemented yet
+fn collect_transitive_requires(
+    main_dependencies: &[String],
+    metadata: &Metadata,
+) -> Result<Vec<String>> {
+    let mut merged_requires = main_dependencies.to_vec();
+
+    // https://peps.python.org/pep-0508/#names
+    // https://peps.python.org/pep-0440/#version-specifiers
+    // This isn't the full version spec regex (not sure if that even exists anywhere),
+    // but it should be good enough
+    let distribution_name_re = Regex::new(
+        r"^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9._-]*[a-zA-Z0-9])\s*(~=|==|!=|<=|>=|<|>|===|;|$).*",
+    )
+    .unwrap();
+    // Since we don't have PEP440 parsing, we do this very simple scheme and to warn the user when we have multiple requires
+    let get_distribution_name = |dep: String| {
+        distribution_name_re
+            .captures(&dep)
+            .with_context(|| {
+                format!(
+                    "Invalid expression (no matching distribution name) in dependencies: `{}`",
+                    dep
+                )
+            })
+            .map(|name| name.get(1).unwrap().as_str().to_string())
+    };
+
+    let mut main_deps_names: HashMap<String, String> = HashMap::new();
+    for dep in main_dependencies {
+        let dist_name = get_distribution_name(dep.clone())
+            .context("Your dependencies table in pyproject.toml is invalid")?;
+        main_deps_names.insert(dist_name, dep.to_string());
+    }
+    debug!("main dependencies {:?}", main_deps_names);
+
+    // distribution name -> (requires string, crate name)
+    let mut metadata_deps_names: HashMap<String, Vec<(String, String)>> = HashMap::new();
+    // I've manually tested that this doesn't include optional dependencies
```

Comment:
But does it include enabled optional dependencies, e.g. if there is an optional dependency on `rust-numpy` but it is in use? (That is rather common in my experience, i.e. writing a Rust library that has Python bindings involving NumPy types behind a feature flag to avoid pulling in PyO3 and rust-numpy unconditionally for the Rust users.)